core/libp2p/transport/tcp/tcp_connection.hpp

```diff
@@ -8,64 +8,171 @@
 
 #define BOOST_ASIO_NO_DEPRECATED
 
-#include "libp2p/transport/connection.hpp"
+#include "libp2p/connection/raw_connection.hpp"
 
 #include <array>
 #include <boost/asio.hpp>
+#include <boost/noncopyable.hpp>
+#include <ufiber/ufiber.hpp>
+#include "libp2p/multi/multiaddress.hpp"
+#include "libp2p/transport/tcp/tcp_util.hpp"
 
 namespace libp2p::transport {
 
   /**
    * @brief boost::asio implementation of TCP connection (socket).
+   * @tparam Executor boost asio executor type. Can be a strand, io context or
+   * custom.
    */
-  class TcpConnection : public Connection,
-                        public std::enable_shared_from_this<TcpConnection> {
+  template <typename Executor = boost::asio::io_context::executor_type>
+  class TcpConnection
```

Comment:
Why have you put everything into a .hpp file? Sounds not great

Reply:
https://github.com/boostorg/asio/blob/develop/example/cpp11/executors/priority_scheduler.cpp#L16

We need a template, because executor may not have a base class.
There are many types of executors - default from io_context, a strand, custom executor. 
Strand is a primitive, which helps to synchronize callback hander execution, so you don't use explicit locks to do that (in multithreaded server). 
Right now we don't use strand, because it is not clear if we need it at all. But from different sources I've read that potentially in multithreaded server we would need to create a strand per connection, so we need to support multiple executors. And since there is no base class, the only way is to make it template